ion/symbols_table.go

```diff
@@ -0,0 +1,124 @@
+	symbolTableKindLocal
+)
+
+// SymbolTable is the core lookup structure for Tokens from their symbolic ID
+type SymbolTable struct {
+	kind    symbolTableKind
+	name    string
+	version int32
+	imports []SymbolTable
+	textMap map[string]SymbolToken
+	tokens  []SymbolToken
+	maxSID  int64
+}
+
+func newSymbolTableRaw(kind symbolTableKind, name string, version int32) *SymbolTable {
+	table := SymbolTable{
+		kind:    kind,
+		name:    name,
+		version: version,
+		textMap: make(map[string]SymbolToken),
+	}
+	return &table
+}
+
+// newSymbolTable constructs a new empty symbol table of the given type
+// Imports are the tables to load into this one.
+func newSymbolTable(kind symbolTableKind, name string, version int32, imports ...SymbolTable) *SymbolTable {
+	table := newSymbolTableRaw(kind, name, version)
+	// copy in the import "references"
+	table.imports = append(imports[:0:0], imports...)
+	if kind == symbolTableKindLocal {
+		// for local symbol table, import system symbol table
+		imports = append([]SymbolTable{systemSymbolTable}, imports...)
+	}
+	for _, importTable := range imports {
+		// TODO Consider if we should not inline the imports or make this delegate to the imports or configurable
+		table.tokens = append(table.tokens, importTable.tokens...)
+		for text, newToken := range importTable.textMap {
+			_, exists := table.textMap[text]
+			if !exists {
+				table.textMap[text] = newToken
+			}
+		}
+		table.maxSID += importTable.maxSID
+	}
+	return table
+}
+
+// newLocalSymbolTable creates an instance with symbolTableKindLocal.
+// The name is empty and the version is zero, these fields are inapplicable to local symbol tables.
+func newLocalSymbolTable(imports ...SymbolTable) *SymbolTable {
+	return newSymbolTable(symbolTableKindLocal, "", 0, imports...)
+}
+
+// newSharedSymbolTable creates an instance with symbolTableKindShared.
+// returns `nil` if the version is not positive.
+func newSharedSymbolTable(name string, version int32, imports ...SymbolTable) *SymbolTable {
+	if version <= 0 {
+		return nil
+	}
+	return newSymbolTable(symbolTableKindShared, name, version, imports...)
+}
+
+// BySID returns the underlying SymbolToken by local ID.
+func (t SymbolTable) BySID(sid int64) (SymbolToken, bool) {
+	if sid <= 0 || sid > t.maxSID {
+		return symbolTokenUndefined, false
+	}
+	return t.tokens[sid-1], true
+}
+
+// ByText returns the underlying SymbolToken by text lookup.
+func (t SymbolTable) ByText(text string) (SymbolToken, bool) {
+	tok, ok := t.textMap[text]
+	if !ok {
+		return symbolTokenUndefined, false
+	}
+	return tok, true
+}
+
+// InternToken adds a symbol to the given table if it does not exist.
+func (t *SymbolTable) InternToken(symText string) SymbolToken {
+	tok, exists := t.textMap[symText]
+	if !exists {
+		t.maxSID += 1
+		var source *ImportSource = nil
+		if t.kind != symbolTableKindLocal {
+			// defining a token within a shared/system table makes the token have a source referring to its
+			// own table and SID
```

Comment:
Nitpick: Sentences start with capital letters and end in periods.

Reply:
Fixed.